aptos-move/framework/supra-framework/sources/automation_registry.move

```diff
@@ -0,0 +1,336 @@
+
+        let automation_registry = borrow_global_mut<AutomationRegistry>(@supra_framework);
+        let ids = enumerable_map::get_map_list(&automation_registry.tasks);
+
+        let current_time = timestamp::now_seconds();
+        let expired_task_gas = 0;
+
+        // Perform clean up and updation of state
+        vector::for_each(ids, |id| {
+            let task = enumerable_map::get_value_mut(&mut automation_registry.tasks, id);
+            if (task.expiry_time < current_time) {
+                expired_task_gas = expired_task_gas + task.max_gas_amount;
+                enumerable_map::remove_value(&mut automation_registry.tasks, id);
+            } else if (!task.is_active && task.expiry_time > current_time) {
+                task.is_active = true;
+            }
+        });
+
+        // Adjust the gas committed for the next epoch by subtracting the gas amount of the expired task
+        automation_registry.gas_committed_for_next_epoch = automation_registry.gas_committed_for_next_epoch - expired_task_gas;
+    }
+
+    /// Withdraw user's deposited fee from the resource account
+    entry fun withdraw_fee(supra_framework: &signer, to: address, amount: u64) acquires AutomationRegistry {
+        system_addresses::assert_supra_framework(supra_framework);
+
+        let automation_registry = borrow_global_mut<AutomationRegistry>(@supra_framework);
+        let resource_signer = account::create_signer_with_capability(
+            &automation_registry.registry_fee_address_signer_cap
+        );
+
+        supra_account::transfer(&resource_signer, to, amount);
+
+        event::emit(FeeWithdrawn { to, amount });
+    }
+
+    /// Update Automation gas limit
+    public entry fun update_automation_gas_limit(
+        supra_framework: &signer,
+        automation_gas_limit: u64
+    ) acquires AutomationRegistry {
+        system_addresses::assert_supra_framework(supra_framework);
+
+        let automation_registry = borrow_global_mut<AutomationRegistry>(@supra_framework);
+        automation_registry.automation_gas_limit = automation_gas_limit;
+
+        event::emit(UpdateAutomationGasLimit { automation_gas_limit });
+    }
+
+    /// Update duration upper limit
+    public entry fun update_duration_upper_limit(
+        supra_framework: &signer,
+        duration_upper_limit: u64
+    ) acquires AutomationRegistry {
+        system_addresses::assert_supra_framework(supra_framework);
+
+        let automation_registry = borrow_global_mut<AutomationRegistry>(@supra_framework);
+        automation_registry.duration_upper_limit = duration_upper_limit;
+
+        event::emit(UpdateDurationUpperLimit { duration_upper_limit });
+    }
+
+    /// Calculate and collect registry charge from user
+    fun collect_from_owner(owner: &signer, _expiry_time: u64, _max_gas_amount: u64) {
+        // todo : calculate and collect pre-paid amount from the user
+        let static_amount = 100000000; // 1 Aptos
```

Comment:
for the time being we can avoid applying dynamic charging, and we can collect flat rate based on the expiry-time and `automation_unit_price`.